### ISSUE TITLE
Feature: CLR Thread Scheduler and Scheduler API (a.k.a. Green Threads)

### ISSUE THREAD

**Issue by RobertBouillon** [dotnet/runtime#50796]
# Overview
Create an abstraction layer that allows the CLR to run its own thread scheduler.

## What problem(s) does this solve?
1. Deprecate complexity and instability created by the async/await pattern
2. Improve performance of context-switching

## Compatability
 No breaking changes to the API. Backwards-compatible with async/await.

## Tasks _(Proposed)_

- [ ] Design abstract API based on existing OS-threading API
- [ ] Proof-of-Concept for CLR-based thread scheduler
- [ ] Design suite of tests to compare async/await against a CLR threading model

# Technical Details
Async/await is an attempt to mitigate the cost of context-switching, but does so by adding significant cost and complexity to software. This proposal is a successor to async/await: an attempt to achieve the same performance gain without the cost (without the language and compiler overhead of the async/await pattern). This proposal is similar to Java's "Green Threads" and Go's "Goroutines."

There is a specific technical challenge that needs to be flushed out in a PoC, first, and that is how to persist _part_ of the stack. During a context switch, the OS swaps out the thread's context, including its stack. This is expensive. If we're performing a user-mode context switch, we only need to persist the stack above our custom thread scheduler. This smaller stack copy is where we gain our performance. Exactly **how** to do this has been explored by other languages, but hasn't necessarily been solved.

## Problems with the current design
Async/await is driven by a form of thread-scheduler that competes directly with the scheduler provided by the OS. While in an "async context", blocking and threading primitives are prohibited. Async/await is prohibited outside of an "async context." Thus, these execution models / contexts are mutually exclusive. Despite being mutually exclusive, it is possible to try to "block" in an async context, or to forcefully await a task in a synchronous context, with potentially catastrophic results. 

In addition to the stability and pattern caveats of async/await, methods also have to be duplicated for both execution models (e.g. `void Foo` and `Task FooAsync`). Code duplication immediately suggests that some form of abstraction would be beneficial, however it is not possible to write methods that would work in both sync and async contexts (The `await` keyword provides necessary metadata for state machine generation). If we can achieve the same benefits of async/await without the caveats, we should, but this depends on our ability to create a high-level thread scheduler.

## API Suggestions
An abstract API should allow operations to be performed against the currently executing context (similar to SycnchronizationContext, but specialized for thread scheduling).

```csharp
//Threading API
static void Foo() => Thread.Sleep(0);

//Async API
static async void FooAsync() => await Task.Yield();

//Proposed Abstract API
static void foo() => Scheduler.Yield();
```

This would allow code to be written agnostic of the threading model: the code doesn't change whether we're using a "green" thread, an OS thread, or an async/await state-machine (presuming the latter is even still possible).

Ideally, most of the classes in the `System.Threading` namespace would be updated, allowing for advanced thread synchronization - something that's currently not possible with async/await.

**Comment by HaloFour** [comment]
> This would allow code to be written agnostic of the threading model: the code doesn't change whether we're using a "green" thread, an OS thread, or an async/await state-machine (presuming the latter is even still possible).

As mentioned in the previous thread, I don't think this would change much with how `async`/`await` works.  At best the state machine would be converted to continuation primitives instead.  That would be nice, and I think those primitives would be useful in other cases, especially if they can be used to build other forms of coroutines.  But beyond that I don't imagine that the runtime would be onboard with trying to make the entire .NET ecosystem safe for green threads or that the C# team would be willing to move away from `async`/`await`.

I'd also argue that green threads are not about mitigating the costs of context switching, they're about mitigating the cost of blocking:  https://inside.java/2020/08/07/loom-performance/

**Comment by CyrusNajmabadi** [comment]
> Async/await is an attempt to mitigate the cost of context-switching, 

I don't think this is the case.  Indeed, i can't recall any discussions around teh design of the features/apis here that brought that into consideration.  Can you clarify this idea more to help ground the motivation of the discussion?

**Comment by CyrusNajmabadi** [comment]
> Async/await is driven by a form of thread-scheduler

`async/await` is orthogonal to threads or other concurrency concerns.  It's a means of expressing asynchrony and abstracting a programming pattern over it, with internal regard to making asynchronous points both evident in code, and non-blocking by default.  It was a design goal of async/await to make such points visually apparent and to not attempt to hide them behind code that looked the same as synchronous threaded code.

> the code doesn't change whether we're using a "green" thread, an OS thread, or an async/await state-machine

This is an drawback to me.  It's an anti-benefit and it would make many forms of development more challenging IMO.

**Comment by RobertBouillon** [comment]
> Can you clarify this idea more to help ground the motivation of the discussion?

I see async/await as a means to avoid blocking. Blocking presents two primary problems - performance (context switches) and deadlocks. I don't see how async/await avoids the latter, so I presume it to be the former.

> ...with internal regard to making asynchronous points both evident in code, and non-blocking by default.

Can you elaborate on this? What specifically makes a method "asynchronous" and why should it be non-blocking? As a developer, how do I benefit from a method being decorated as asynchronous? Why do I care if it blocks if I depend on its successful execution, anyway?

To be more specific, consider the following code:
```csharp
int Download(string url) => Int32.Parse(WebClient.Download(url));

int async DownloadAsync(string url) =>Int32.Parse(await WebClient.DownloadAsync(url));
```
Both methods do the exact same thing. How does the extra syntax required for the async version benefit me? 

I presumed this to be a performance optimization. If it's not, I see less of a reason for it to exist. I worry I'm missing something, but every time I come back to it I see traditional threading as simpler / superior to async/await.

**Comment by benaadams** [comment]
> Both methods do the exact same thing. How does the extra syntax required for the async version benefit me?

The `async` one releases the thread of the scheduler (whether UI thread, ThreadPool thread or other) to do other work while its waiting for the download result.

The first one will lock the thread until it gets the result; freezing the UI or taking the ThreadPool thread out of circulation for no useful work. 

On the default scheduler (ThreadPool) for a non-UI app; `async`/`await` is already essentially acting as green threads?

**Comment by CyrusNajmabadi** [comment]
> Both methods do the exact same thing. How does the extra syntax required for the async version benefit me?

One releases the thread to do other things, the other does not. It definitely is a decision I care about when writing a program.  I don't want threads going off and doing other with unless I tell it to do that.

**Comment by CyrusNajmabadi** [comment]
>Can you elaborate on this? What specifically makes a method "asynchronous" 

It returns something 'task like'. 

This means is represents a computation that can complete sometime in the future, not when the method returns.

**Comment by HaloFour** [comment]
> every time I come back to it I see traditional threading as simpler / superior to async/await.

It's not a performance optimization, it's a resource optimization.  Threads are expensive and must be paid for upfront in preallocated memory for their stack.  Blocking them on I/O creates a bottleneck in the number of concurrent operations you can have in flight.

**Comment by RobertBouillon** [comment]
There are a couple of ways to look at my example:

From an OOP/API perspective, the function `Download` encapsulates two operations - a download and a parse. As a developer, I've successfully expressed that the parse depends on the download. "How" the thread is suspended while waiting for the download to complete is outside of the scope of the function - it needs not be defined in the body or the signature, nor should it be. We can call this "Separation of Concerns," but I think it's more fundamental OOP - the method signature is independent of the threading model. The method body should define a sequence of operations. Async/await does not add value here - only complexity.

Perhaps more important is the execution: in both cases a sequence of operations depends on external data. In both cases, the state of the currently executing thread is stored and the hardware thread is redirected to another sequence of operations until the data is available. When the data is available, the original sequence is resumed with the newly available data. Async/await is just another way of keeping the hardware thread busy, with the same objective as the OS thread scheduler, only executed in different ways. 

So if we were always able to suspend and resume a thread via traditional threading, what's the benefit of async/await?

> The async one releases the thread of the scheduler (whether UI thread, ThreadPool thread or other) to do other work while its waiting for the download result.

This needs to be broken down to better understand my argument. Simplified for argument's sake, a hardware thread is the capacity to execute instructions (i.e. A CPU core). A software thread is a hardware thread and a stack allocation. When you say that a thread is "released to do other work", what you're saying is that the stack is popped down to a common location so it can be reused. This is done to prevent the allocation of another software thread (or the restoration of a suspended one). So this is an optimization.

Whenever I trace the "why" for async/await down to bedrock, I always find optimization to be the underlying reason, specifically with respect to thread context. This being the case, the best way to optimize thread context would be to expressly manage contexts in the CLR. If you need to hit a nail, you find a hammer. Async/await places a significant burden on the developer and API for something that seems to be the domain of the CLR.

> On the default scheduler (ThreadPool) for a non-UI app; async/await is already essentially acting as green threads?

Not really because you can't use thread synchronization primitives with async/await (they block). Green threads would restore the ability to use thread synchronization for managing critical sections. Async/await also requires the use of async/await decorations, where green threads could operate without the need for specific method signatures or syntax decorations. A synchronous method would be identical to an asynchronous one. Execution context would always be explicit.

> I don't want threads going off and doing other with unless I tell it to do that.

I agree. For example, I don't want my method's execution to be arbitrarily assigned to a thread in the thread pool without my control, which is what async/await does. When I call a method, I expect the method to be executed sequentially in the current context (the current thread). In the rare cases where I want parallelism, I want to explicitly manage the context of that parallelism. I don't care that a thread _could_ run asynchronously (any method _could_) - I'm calling it because I want it to run synchronously. If I wanted it to run asynchronously, I'd set up a new context (New thread or pooled thread) and call the method from there. Returning a `Task` seems unnecessary because most of the time I'm just going to `await` it, and when I don't, I expect to have to do some work to manage the parallelism. Returning a `Task` doesn't make managing the parallelism any easier.

**Comment by HaloFour** [comment]
Green threads and `async`/`await` coroutines are very similar constructs.  Both are continuations that capture the execution state within the method.  Both allow that continuation to be resumed to continue that execution.  Both require that the _entire_ ecosystem around them be redesigned to support them.

The goal of green threads is to make "blocking" cheap, but they can't do this automatically.  They can only accomplish this if no method you call actually blocks.  Every single one of those potentially-blocking methods has to be written to recognize that they are running within the context of a green thread, to capture the execution state of that green thread into a continuation, to use some kind of notification mechanism to trigger resuming that continuation and to return the OS thread back to the green thread scheduler.  If you block a green thread you actually block the underlying OS thread, which brings you right back to square one, except worse as those potentially millions of green threads can only be scheduled by a much smaller number of OS threads.  It's quite literally impossible to avoid this.

Green threads doesn't achieve anything that you're suggesting, it's just a different strategy for hiding the fact that you're moving continuations between different OS threads.  It's an exceptionally heavy lift on the part of the ecosystem because quite literally every part of it needs to play along.  And none of that changes how it works in the language.

**Comment by CyrusNajmabadi** [comment]
> the method signature is independent of the threading model. 

I don't view this as a good thing. If downloading and parsing may happen asynchronously, I want to know so I can plan accordingly from the code that calls it.

Perhaps the caller will choose that that's ok, and it sounds just block until that work is done. That's fine, it can make that choice. Perhaps it will also want to yield the thread of execution. If so, that's fine. Perhaps it wants to change the state of the system while that async work is happening, and change it again once complete. That's fine, it can do that.  Perhaps it wants to poll for the result for some amount of spins, and then change it's strategy after a while. That's fine, it can do that. Perhaps it will decide that it would like to do something else while that is happening, perhaps explicitly concurrently. It can do that too. Etc, etc.  

By handing back the entity that allows one to represent the "work that will complete at some point in the future", this code can easily do all of this, as well as easily compose over N other pieces of with like this. You bring up OO, and I would say this is the OO way if representing 'a computation in flight'. I want that to be explicit and easy to work with.

**Comment by HaloFour** [comment]
> I don't want my method's execution to be arbitrarily assigned to a thread in the thread pool without my control

This is what happens with green threads.  You don't suspend and resume a thread.  You suspend and resume a continuation.  That continuation is the "green thread", the captured stack space of the thread at the point it was suspended.  When you resume the continuation it will be on whatever OS thread is available, which will almost certainly be a different OS thread than you originally suspended.  This will certainly matter in cases of UI apps where Win32 knows nothing about the green threads and will fail if the underlying OS thread is not the same as the UI thread.

**Comment by RobertBouillon** [comment]
> By handing back the entity that allows one to represent the "work that will complete at some point in the future"

Can't we do this already with `Task.Run`? Why does this belong in the method signature? This seems like the responsibility of the caller, not the callee. Do you have an example where returning `Task` provides functionality that is otherwise not possible? Where the callee must manage its call context?

> I will often want to do something special there (like update ui state). 

In the case of a UI, you don't want the UI thread to do **anything** except rendering the UI, anyway. Any non-trivial method would belong on a background thread. We can't say that returning a task indicates that the method is non-trivial because it's entirely possible to code a non-trivial method that _doesn't_ block and _doesn't_ return a `Task`. If the concern is whether or not a method is long-running, returning `Task` doesn't seem to be the appropriate solution. 

@HaloFour 
Take away async/await. Forget backwards-compatibility concerns. For the sake of argument, async/await never existed. What have we lost, exactly? What is no longer possible? What is objectively worse or more difficult?

**Comment by CyrusNajmabadi** [comment]
>Can't we do this already with Task.Run

No.  Task.Run will place the work on the threadpool.  Furthermore, it forces me to have to somehow know to do this, instead of hte method actually letting me know that this is what is being done.  As the *consumer* of any API, i very much want to know if this will happen so i can plan accordingly at the callsite.

>Why does this belong in the method signature? 

So that i know this is what is going on at all consumption sites.

> This seems like the responsibility of the caller, not the callee. 

I cannot take responsibility here unless my callee let's me know this is going on.  Furthermore, i don't want them to tell me that this won't happen in one release, and then under the covers they change their impl such that this happens in the future.  If this happens in the future, i want their signature to change as i must adapt and react to that new approach they are taking.

> Any non-trivial method would belong on a background thread.

How do i know it's non-trivial?  If the method signature tells me nothing, literally anything could be non-trivial.  This *is* the reason we went this way, and it's why things like the windows coding guidelines explicitly call out that non-trivial methods have these signatures.  It's precisely so that this information is communicated through teh type system, and cannot be ignored.  

> We can't say that returning a task indicates that the method is non-trivial because it's entirely possible to code a non-trivial method that doesn't block and doesn't return a Task

THis would violate the coding guidelines for the platform.  Conversely though, if you do take a lot of time, you should use Task as part of those guidelines, and now your callers need to deal with that.  In a green thread world.  If you take a lot of time... then what?  How do i, as the caller, know that?   If you initially dont' take a lot of time, and later on change to take a lot of time, how do i know as a caller that that's the case?  How do i appropriately break so that i must react to this?

> What is objectively worse or more difficult?

It is more difficult to write responsive apps.  It is more difficult to understand what individual lines and blocks of code are doing and if they will have a problematic impact on the app.  It is more difficult to ensure that changes to underlying libraries do not negatively impact your app.

**Comment by HaloFour** [comment]
@RobertBouillon 

> Take away async/await. Forget backwards-compatibility concerns. For the sake of argument, async/await never existed. What have we lost, exactly? What is no longer possible? What is objectively worse or more difficult?

1. You can't forget those concerns.  We're not replacing .NET, or Win32, or Mono, or Xamarin, or the millions of lines of third party native code that runs on the .NET framework, or the billions of lines of native code that can be called by the .NET framework.  Whatever solution here _must_ work in the existing ecosystem.
2. Without `async`/`await` you end up where you were with TPL and where most async frameworks in most languages are without coroutines, composition via callback methods.  Having been there I can state unequivocally that it is substantially worse.  This is why `async`/`await` is becoming so popular across so many languages.

**Comment by RobertBouillon** [comment]
> As the consumer of any API, i very much want to know if this will happen so i can plan accordingly at the callsite

I'm not following some of what you're saying when you say "this" will happen. What is "this"? I want to be more specific. What is happening that you seek to capture when `Task` is returned? That the operation / thread has been suspended?

> THis would violate the coding guidelines for the platform. 
Could you link these guidelines? (I looked and couldn't find - I apologize). 
So I have an MD5 operation in a tight loop. `Stream` in, `Stream` out. You're saying this needs to be rewritten to return a `Task<Stream>` to conform to guidelines?

**Comment by jnm2** [comment]
> From an OOP/API perspective, the function Download encapsulates two operations - a download and a parse. As a developer, I've successfully expressed that the parse depends on the download. "How" the thread is suspended while waiting for the download to complete is outside of the scope of the function - it needs not be defined in the body or the signature, nor should it be.

From a conceptual perspective, returning `Task<int>` immediately gives the caller a much richer set of possibilities than blocking until the work is finished and returning `int`.

**Comment by CyrusNajmabadi** [comment]
> I'm not following some of what you're saying when you say "this" will happen. What is "this"?

**Comment by RobertBouillon** [comment]
Is this a good time to ask for clear, well-maintained guidelines?

If you're saying that I need to return a `Task` for any method that _could_ take longer than 50ms, that's absurd. It's absurd because execution time could be dependent on the environment or the input. It means as I develop my software, my API would need to change based on its behavior characteristics (and dependent code refactored).  It means I have to wrap synchronous code into a task **_just-in-case_** the caller wants to make it asynchronous because if it _might_ take longer than 50ms. It's absurd because "a long time" is completely subjective. It means I can't create completely abstract interfaces because they depend on implementation. It violates mature and well-established OO principles. I hope this notion remains dead and buried in an old blog because it's 72 flavors of wrong.

I created this presuming that the reason for async/await was performance. What I'm hearing is that async/await facilitates the explicit desire to label methods as "long-running" and manage parallelism around them. Is this correct? If not, could you rephrase it succinctly for me so I can better understand it? And could you clarify if this reflects the .NET team's design objectives? I'm not clear on who's who in this discussion.

**Comment by HaloFour** [comment]
I think that the official docs do a fairly good job of describing when and why you'd use `async`/`await`:

https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/

**Comment by CyrusNajmabadi** [comment]
> Is this a good time to ask for clear, well-maintained guidelines?

**Comment by CyrusNajmabadi** [comment]
> If you're saying that I need to return a Task for any method that could take longer than 50ms, that's absurd.

I don't think it's absurd.  

>  or the input.

Yes.  if an input could really cause things to make much more time, that really should be encoded into your sig.  Because your caller needs to adapt to that.  I don't want to click a button, and have my entire UI freeze because i pass a filename to you that you end up taking 50 seconds on.   Yes, you may tke 10ms on some files.  But if if you are going to realistically take that long, then i need to know so i can design my own code accordingly.

>  It means I can't create completely abstract interfaces because they depend on implementation

Then use Task in those interfaces.  It accurately represents that impls may take a long time.

Note: this is not theoretical.  I'm currently deprecating an existing Roslyn api that operates synchronously because we need it to be async.  This *does* impact all consumers of this api and they *do* need to adjust because they should no longer assuem that this code is fast.

> It's absurd because "a long time" is completely subjective

**Comment by CyrusNajmabadi** [comment]
> And could you clarify if this reflects the .NET team's design objectives? I'm not clear on who's who in this discussion.

I'm a member of the C# language design team.  I was heavily involved in the design around async/await and the general principles we had in mind when designing it.  What i'm describing is one, of many, considerations taht were considered important when we were designing this.

It was a design *goal* to make asynchrony apparent, and to make it easy to compose operations over them.

> What I'm hearing is that async/await facilitates the explicit desire to label methods as "long-running" and manage parallelism around them

I would not use the work parallelism here necessarily.  But 'composing asynchrony' was certainly a goal.

**Comment by RobertBouillon** [comment]
I appreciate the time. So help me out here -

Code, by its nature, is synchronous; code is a predefined sequence of operations. Functions are also inherently synchronous, with a single entry and exit-point - functions are fundamentally a named sequence of operations with a well-defined input and output. What breaks the simplicity is the concept of a "blocking operation." A method may cause the sequence of operations to become idle, locking up resources associated with the thread. 

My understanding of async/await is that, instead of blocking, we now return a `Task`. This gives the caller the option to suspend execution until the method completes or continue the execution in some way. I find asynchrony to be a misnomer here because what we're really talking about is parallelism. We either `await` the task, preserving the code's synchronous execution, or we perform other operations, inducing parallelism. Just to be clear, when I say "preserving the code's synchronous execution", I understand that it's technically yielding instead of blocking. For all intents and purposes, though, `await` exists to provide the illusion of synchronous execution, the same as blocking does.

So when we talk about async/await, we're still talking about parallelism. I've seen this suggested a few times, and even in here, that async/await is orthogonal to threading, and maybe that's true conceptually, but not in practice or as-implemented. Async/await has a hard dependency on a thread scheduler that's defaulted to the thread pool, and it prevents the use of threading primitives. It's more appropriate to say async/await is mutually exclusive to threading, since you no longer have direct control over thread context. You're either pulling your own threads in a sync context or running in an async context, but mixing the two is inadvisable (and against most [best practices](https://docs.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming)).

As far as I can tell, async/await is just another way to fork a process. With async/await, any time a `Task` is returned, you can optionally execute parallel code. With threading, you must explicitly invoke a parallel process. I would argue that threading is superior for many reasons: I have access to advanced and mature synchronization primitives, I have finely-tuned control over thread execution and thread context, and my code executes as-written with no state machines or performance hits from hidden boxing / unboxing. I also don't have to deal with the [two dozen gotchas](https://docs.microsoft.com/en-us/archive/msdn-magazine/2015/july/async-programming-brownfield-async-development) [because of async/await voodoo](https://docs.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming). Oh, and my abstract interfaces don't need two different versions of every method (one sync and one async) to be complete. Threading gives me options - async/await gives me restrictions.

So I find myself pursuing our conversation in search of answers again - that maybe you've already answered? What's the benefit of async/await?

If the goal was to make "asynchrony apparent" or facilitate "asynchrony composition," how was it not possible already with traditional threading? If a `WebClient.Download` blocks while waiting for a result, there's no asynchrony to be aware of - it's a synchronous operation that's blocking for a result.  If I need to do something else while it downloads, I'd put it on another thread. In the case of `IAsyncResult WebClient.DownloadAsync`, asynchrony was still apparent. There's no obfuscation for async/await to demystify. 

Async/await makes sense, conceptually, as an evolution of `IAsyncResult`, since it replaces callbacks with a hidden state machine. That's cool, and that's where it should end. The fact that async is then needed "all the way down" increases the cost way beyond the benefit.

So _maybe_ I can see await/async for performance-critical I/O operations that return a low-level handle, but "any operation that could be long-running"? That may as well be everything.

**Comment by HaloFour** [comment]
https://blog.stephencleary.com/2013/11/there-is-no-thread.html

**Comment by HaloFour** [comment]
> If I need to do something else while it downloads, I'd put it on another thread.

Which is a giant waste of resources.  Threads are expensive, which is why we end up with pools.  Blocking threadpool threads leads to exhaustion.   Using threads for the sake of preventing an I/O operation from blocking is totally unnecessary when most I/O operations have OS-level facilities to notify on completion.  And creating threads is the easy part, synchronization of the result is complicated.  Coroutines like `async`/`await` make this kind of development orders of magnitude simpler.  Nothing can be simpler than single threaded synchronous blocking code, but little can be as wasteful.  Server-side it tanks throughput.  Client-side it renders the UI totally unresponsive.  At this point I don't know what the argument is here.

**Comment by RobertBouillon** [comment]
>At this point I don't know what the argument is here.

I'll tell you what my objectives are:
1. I want to use threading primitives in my code. Invoking a coroutine shouldn't prevent this. Async/await does. I can do advanced threading or async/await, but not both. These should not be mutually exclusive.
2. I don't want to write the same method twice - [once for async](https://github.com/dotnet/runtime/blob/5bad843281caac09e18958f27d7d914eb6086e77/src/libraries/System.IO.FileSystem/src/System/IO/File.cs#L937-L972) and [once for sync](https://github.com/dotnet/runtime/blob/5bad843281caac09e18958f27d7d914eb6086e77/src/libraries/System.IO.FileSystem/src/System/IO/File.cs#L492-L526). Nor define both in my abstractions. Writing a function once should be enough, and always has been before async/await. My method should not care how it's being executed, nor should I care when I'm writing it. This is code duplication at its worst and there must be a better way. If the root problem is that blocking is inefficient, that sounds like a CLR problem, not a syntax problem.
3. I don't want my code to be split between "sync" and "async" with the software-equivalent of a minefield bridging the two. Any serious foray into an async application has made me loathe C# for making complex what should be simple. This is the exact opposite of my experience with C# to this point.

I'm trying to understand how to accomplish the same thing as async/await without the massive headache it brings. My thought was green threads. If you're telling me that's crazy, help me to understand what isn't. Whatever problems async/await solves can't be worse that the problems it creates. There's no way this is "the best way," because at least in my experience, code was way cleaner before async/await, and I've written more than my fair share of asynchronous code.

**Comment by CyrusNajmabadi** [comment]
> I want to use threading primitives in my code. Invoking a coroutine shouldn't prevent this. Async/await does. I can do advanced threading or async/await, but not both.

I don't really understand this claim. Nothing and async/await prevents threading (and I'm not quite sure that 'advanced' refers to here). Async/await works quite complimentary with threading. 

> Any serious foray into an async application has made me loathe C# for making complex what should be simple.

I don't really understand this either. Using async has made things so much nicer for me in terms of app development. The code is clear at the points when this work happens, and I don't have to care at all about blocking my ui, or sharing state inappropriately across threads. 

> without the massive headache it brings

I think the disconnect is here. You view this as a headache. From reading your posts it's seems that this stems from you trying to hide this so you don't have to think about it. But the system saying: it does really matter, think carefully about what you're doing. 

That said, of you really want to just avoid async, you can do so. Just do a blocking read wherever this pops up. It is highly recommended that you not do this. But if you'd prefer to not have the asychrony, then this is the natural fallout here.

>  code was way cleaner before async/await

It definitely was not. Take a simple example of any sort of loop that has to operate on something asynchronously.  Or nested loops that need to communicate data to/from each other. This would involve heinous callback soup to try to switch together all the async work into something sensible. You'd also get the towers of doom that arise with lots of nested callbacks. 

With asyn/await I get serial logic, but without the problems that that brings when doing when synchronously.

**Comment by CyrusNajmabadi** [comment]
> I'm trying to understand how to accomplish the same thing as async/await without 

I think a core issue here is that the *only* story .Net is backing here is 'async'.  No efforts are being made toward any other styles, and the entire ecosystem has rallied behind this. So if you push against that you'll likely find nothing but headaches, and likely very little interest or engagement towards anything else. From the language down to the apis and ecosystem, this is the pattern that was settled on. 

It would be akin to saying: I want to use .net, but I don't want to call methods.  Sure, you might find some way to accomplish it, but you'll be fighting upstream every step of the way.

**Comment by HaloFour** [comment]
@RobertBouillon 

I think I can understand that desire and I think it's totally reasonable.  Unfortunately there really aren't any good answers here, regardless of the language or framework.  At best you can ignore asynchronous programming and continue to use imperative blocking and threading where you deem it appropriate.  When dealing with async APIs you can compose over the task manually, or just block on the result via `GetAwaiter.Result`, but you'll have to be wary about deadlocks if working with a framework that necessitates a synchronization context.  When it comes to any asynchronous programming style, whether that be green threads, coroutines, promises, callbacks, etc., you'll be bouncing between threads, so synchronization primitives that are designed with thread affinity won't work.  In those languages that have green threads they've had to design/rewrite all of their synchronization primitives specifically so that they don't care about the thread but only about the identity of the green thread.

Asynchronous programming is a very tricky space.  `async`/`await` tries to make it easier for developers and I think succeeds in doing so, compared to the alternatives.  But it's still a tricky space and some of those thread-specific tricks just won't work well in it.  What might find more traction in these forums is opening discussions around where there is friction around synchronization primitives and `async`/`await` as many new flavors of those primitives can be designed that are friendly to these coroutines.

**Comment by RobertBouillon** [comment]
Perhaps it will help if I provide examples.

> Nothing and async/await prevents threading

Most threading primitives leverage locking. Locking is prohibited in async/await.

https://blog.cdemi.io/async-waiting-inside-c-sharp-locks/
https://stackoverflow.com/questions/19659387/readerwriterlockslim-and-async-await

For example, I had an ASP.NET site that required a JWT security token to access network resources. It expired every few minutes and needed to be refreshed. I created a DI service to periodically fetch a new token. Using traditional threading caused deadlocks and there was no clear / easy way to secure the critical section. I ended up writing a custom Spin Lock using `Task.Yield`. Outside of async/await, this was trivial with a ReaderWriterLockSlim. This underpins my request in this issue for an abstract API for threading primitives, and is just one of many examples where async/await has made complex a task that would have otherwise been simple.

> it does really matter, think carefully about what you're doing.

Does it _really_ matter, though? Why didn't it matter _before_ async/await? My ASP.NET sites scaled fine when my ADO.NET calls were synchronous. 

> code was way cleaner before async/await 
> It definitely was not. 

-_-

I have a database engine that leverages GPGPU. It branches out to C++ quite a lot. Many of the low-level interfaces were synchronous, and in-fact, the entire managed API is synchronous. The managed API is a complex wrapper around a native API. We wanted to move to [asynchronous data copies](https://developer.nvidia.com/blog/how-overlap-data-transfers-cuda-cc/) to improve performance, and this seemed like the perfect time to look at async for the managed API, as well. "Async or bust!" Yeah, it was bust. We abandoned the branch. 

We went into the effort accepting the additional complexity of method pairs. Keep in mind that we're talking hundreds of members [in one class alone](https://gist.github.com/RobertBouillon/e57de0beb365401d738ccd3e0de6afbc). We expected the public API to change, but we underestimated the impact on private or "support" classes. So many methods needed to be async that it raised performance concerns we never even vet because the endeavor was abandoned.

We had all types of issues that I couldn't even list them all. Like, some code leveraged expression trees emit at runtime, which were no longer possible because expression trees can't emit the state machine necessary to support async/await. We're going to try an `IAsyncResult` pattern once we're done licking our wounds. It's simpler, more predictable, works with expression trees, and it's analogous to the underlying C++ API. It seems like a more natural fit and a much lower development risk.

> So if you push against that you'll likely find nothing but headaches

Sure, but will the headache be less than those which async/await gives me? :) All kidding aside, I just want to make things better. I'm not a programmer - I'm a problem solver. Code is just my tool of choice. When I see a problem, I try to solve it. Async/await gives me many problems I want to solve.

@HaloFour 
Makes sense. I really wish I could avoid async/await, but alas, I cannot, for the same reasons you mentioned.


I hope that this illustrates a little that my disdain for async/await is not rooted in ignorance of the feature, but in experience with it. Maybe green threads aren't the right approach, I don't know. What I do know is that I could write a book on how async/await has made C# a more painful coding experience, overall. Async C# is like learning a whole new language where someone wanted all the prose of C# and all the pitfalls of JavaScript.

**Comment by CyrusNajmabadi** [comment]
>      Most threading primitives leverage locking. Locking is prohibited in async/await.

Locking is not prohibited in async/await.  We do tons of different forms of threading work in roslyn and intermix it with async functions that use `await`.  Here's an example from just a piece of code i was working on today: 

https://github.com/dotnet/roslyn/blob/98530991561773ac1fbc2c511f40eab211a11790/src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs#L214-L218

That's just hte straight vanilla C# `lock` statement (which is just shorthand for try/finally with some Monitor.XXX calls). But virtually all other locking constructs are available and can be used in a composable fashion.

**Comment by CyrusNajmabadi** [comment]
> Using traditional threading caused deadlocks and there was no clear / easy way to secure the critical section

I honestly don't know why this would be an issue.  I'm not sure hwat 'secure the critical section' means, so it's hard to comment on this part.

>Sure, but will the headache be less than those which async/await gives me?

Yes.  Unquestionably.

> Yeah, it was bust. 

These anecdotes aren't really helpful, because they do not explain what actually went wrong.  So you wanted to move to async data copies.  What *actually* went wrong here? You are blaming async/await for not being able to work here.  But I've seen no information to understand what the actual limitation was.

> My ASP.NET sites scaled fine

Emphasis on "My".  Async/await was about a design to allow effective async program code from all the way from mobile, to desktop, to the server.  You might have been fine for your singular case, but that isn't meaningful when the ecosystem ws *not* fine across a myriad of use cases.  

> Does it really matter, though? 

*Yes*.  very much so.  We're dealing withi this in VS today, which still has old cruft sync apis that are in use (often for legacy plugin apis) that constantly cause anything from microstutters, to full on ghosting in VS.  Having a programming model that can allow this to be simply and effectively avoided is of substantial importance.

> It seems like a more natural fit and a much lower development risk.

IAsyncResult has been a part of the async/await story since its inception.  It's *very* normal to use that pattern to model some systems, and then to abstract over those with the right shapes to enable the standard async/await coding style.  See https://docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/interop-with-other-asynchronous-patterns-and-types for more information on this.  

>  So many methods needed to be async that it raised performance concerns we never even vet because the endeavor was abandoned.

I think a core challenge here is that you took an existing system and attempted to convert it all at once.  When we adopted async/awiat for Roslyn itself, we started small.  We experimented and prototyped.  And we used that time to learn deeply how things worked as well a teach the team the lessons here.  Once we felt comfortable with things and we'd built the right mental and code models, we expanded on this further to the point that it is extensively used throughout the roslyn codebase (practically all places where needed, except for fringe legacy points where we have to talk to synchronous systems out of our control).

**Comment by davidfowl** [comment]
I think doing a green threads implementation would be a worthwhile interesting experiment to do in runtimelab. It's difficult, far reaching and requires work at multiple layers in the stack. I'd file the issue there and discuss what it would actually take in practice (it's much harder than most think when they file issues like this and point to other languages). .NET has constraints other runtimes may not have because of the features we expose. 

Anyways, that said, it's not something trivial that can be done with the list of bullets here.

**Comment by CyrusNajmabadi** [comment]
We even have this extension to make that nicer in roslyn: :)

https://github.com/dotnet/roslyn/blob/98530991561773ac1fbc2c511f40eab211a11790/src/Compilers/Core/Portable/InternalUtilities/SemaphoreSlimExtensions.cs#L19-L24

This way you can just do:

```c#
using var _ = await semaphoreSlim.WaitAsync();
// do work
```

No need for the try/finally/Release bits :)

**Comment by HaloFour** [comment]
@benaadams 

C++ got async coroutines (`co_await`) in C++20, so one more on that list.

AFAICT Java Loom won't run under virtual threads by default, so there would likely need to be some modest user code changes at least during bootstrap.  I imagine frameworks like Spring will likely handle that under the hood possibly enabled by config.

@davidfowl 

I think it'd be awesome if the CLR offered an API for capturing/resuming a continuation. As with Loom I think that would be the first step towards exploring green threads and also offers primitives that would make it substantially easier for languages (and perhaps libraries) to add coroutines without having to build their own state machines.

**Comment by jnm2** [comment]
> For example, I had an ASP.NET site that required a JWT security token to access network resources. It expired every few minutes and needed to be refreshed. I created a DI service to periodically fetch a new token. Using traditional threading caused deadlocks and there was no clear / easy way to secure the critical section. I ended up writing a custom Spin Lock using Task.Yield. Outside of async/await, this was trivial with a ReaderWriterLockSlim. This underpins my request in this issue for an abstract API for threading primitives, and is just one of many examples where async/await has made complex a task that would have otherwise been simple.

I've been in a very similar scenario and made it async all the way down. It's not the most obvious to implement, but it is possible and it is the most true to the reality of what is happening.

**Comment by RobertBouillon** [comment]
> Use `SemaphoreSlim`?

Any thoughts on making thread synchronization more consistent between sync and async code?

More of an observation than a question I immediately care about - why would I use `SemaphoreSlim` instead of `lock`? My access pattern warranted a Reader/Writer Lock, which isn't available in synchronous code because the current implementation is thread-affine. Another question I have that I don't immediately care about is - if I should avoid blocking in an async method, why is it okay to use `lock`, as @CyrusNajmabadi has pointed out? Does this mean I can also use wait handles? Is the concern really just rooted in thread pool exhaustion? 

C# has always been easy for many of the reasons a good UI is easy - because the design is intuitive. If I needed a Reader Writer Lock, there was a class by that name in the Threading namespace. I find async has made C# unintuitive - many of the resources that were natural fits or otherwise standard practice don't apply when coding asynchronously. Is there a way that can be addressed?

My first thought was abstraction - which is why I mentioned it in this issue. Perhaps provide a common API that provides threading primitives that work in both contexts? Maybe a stop-gap would be an MSDN article on thread synchronization in async? Some analysis and guidance on threading primitives in async could provide a blueprint for a more complete & intuitive solution. [Stephen Cleary has something of use here](https://github.com/StephenCleary/AsyncEx) - could / should this be integrated into the BCL?

**Comment by CyrusNajmabadi** [comment]
> More of an observation than a question I immediately care about - why would I use SemaphoreSlim instead of lock? 

They have different semantics. SemaphoreSlim provides semaphore semantics. `lock` provides Monitor semantics. 

Examples of where this matters are things like monitor allowing reentrance for the same thread, while a semaphore does not care about threads and will let anything enter as long as the count is there (or will block of it is not).

**Comment by CyrusNajmabadi** [comment]
> if I should avoid blocking in an async method, why is it okay to use lock

The question implies a disparity where none exists. :-) There is no rule or restriction about locking in async code. Use of synchronization primitives to protect shared state is completely fine. 

What you'd likely want to avoid is *unbounded* synchronous blocking. Same as if I had a sync system. Here, the lock acts to just allow the invariant that both of these are updated at the same time, and adjust provides the right barriers so that the data is consistent between the writing thread and the reading thread.

**Comment by CyrusNajmabadi** [comment]
> If I needed a Reader Writer Lock, there was a class by that name in the Threading namespace.

That type still works, even in async code. Though I find in practice it's not really necessary as it really is a solution for when you view your work as being served by many threads, am if which are free and are running concurrently. 

In a world where the primitive is Task, the questions more naturally become: "when is this running?" And "how to I share data?". For both of these I find there are a lot more powerful abstractions. 

For example, ConcurrentExclusiveSchedulerPair provides the basis for a system where you may want many tasks of one sort in flight, while restricting to only running one task at a time if another sort. 

Similarly, Channels add a great way to pass data along, while decoupling the production and consumption side, but also allowing them to influence each other through things like back pressure. 

All the old threading stuff is still there and still works. Indeed, if threads are the right metaphor for you, you can use them. Tasks translate nicely to this as you can just represent that async, bg with your threads are doing as Tasks.

**Comment by CyrusNajmabadi** [comment]
> many of the resources that were natural fits or otherwise standard practice don't apply when coding asynchronously. 

This is a very broad statement, without specifics.  It's hard to come up with solutions there, because the problems haven't really been explicitly enumerated. For example, several of the issues listed so far (like not being allowed to lock), aren't actually problems, but perhaps stem from a misinterpretation of information somewhere down the line. If you can provide more concrete examples, that would be helpful. Thanks! :-)

**Comment by CyrusNajmabadi** [comment]
> Maybe a stop-gap would be an MSDN article on thread synchronization in async?

Whole I think a lot has been written on this topic (especially in the early days of async), I think more information (esp around solving real issues) can always be helpful. 

I will say this: I'm constantly amazed and enthusiastic about how good the abstractions are here for so many cases. Numerous times over the last 5 years I've come to the runtime to query if there is some api that would make a current async job easier, to find out that they had a rich, will tested, family of apis in that space ready to go for me :-)

The main thing remaining now is just an async version of Lazy (nudge nudge @davidfowl ).

**Comment by RobertBouillon** [comment]
> There is no rule or restriction about locking in async code

There isn't anything about _locking_, but there is about _blocking_, and the `Monitor` class emit `lock` blocks.  [From MSDN: "Don't block, Await instead"](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/#dont-block-await-instead). So if a method blocks, replace it with a method that returns `Task` instead, and `await` it. `Monitor.Enter` blocks. `WaitHandle.WaitOne` blocks. Neither have Async counterparts I can use instead. Many thread synchronization classes lack Async methods, so when I run into classes that have no Async methods, the logical conclusion I draw is that they are not supported and I should not use them in an async context. Even _bounded_ blocking can cause thread pool exhaustion with enough contention, so despite your advice to the contrary, I don't see blocking as an option here, even _bounded_ blocking.

@benaadams example with `semaphoreSlim.WaitAsync` seems to be the appropriate way to effectively `lock` in an async context, so I'll use that moving forward, but there are still many synchronization classes that lack async members.

> [Reader Writer Lock] still works, even in async code

No, it [really doesn't](https://gist.github.com/RobertBouillon/4f3001eb7ee02b2abd8d4762530e27c3).

> Though I find in practice it's not really necessary as it really is a solution for when you view your work as being served by many threads, am if which are free and are running concurrently.

In my use-case, where a REST API servicing thousands of requests per minute tries to access a shared variable that only changes once every 5 minutes, a Reader/Writer Lock was appropriate. And since I couldn't use it, I was forced to use a mutex, instead, which is far less efficient. And worth restating that outside of an async context, using a Reader / Writer Lock is **very** simple and intuitive. 

> because the problems haven't really been explicitly enumerated

Does [this](https://github.com/StephenCleary/AsyncEx) help?


In the interest of "explicitly enumerating my problems" henceforth:
1. Should `lock` emit `semaphoreSlim.WaitAsync` instead of `Montor.Enter` while async?
2. Many thread synchronization classes lack "Async" methods. Is this a gap, and should an issue be created? Or must we rely on [third party libraries](https://github.com/StephenCleary/AsyncEx)? (This [zombie issue](https://github.com/dotnet/runtime/issues/35962) is a similar ask)
3. [ReaderWriterLockSlim does not work in async](https://www.aashishkoirala.com/blog/reader-writer-locking-with-async-await). Is this a bug that should be logged?
4. What is the best way to close this issue? @davidfowl seems to support some effort in runtimelab. Shall I create a new issue? This one is full of contentious digressions...

**Comment by davidfowl** [comment]
> What is the best way to close this issue? @davidfowl seems to support some effort in runtimelab. Shall I create a new issue? This one is full of contentious digressions...

I've been discussing this with a few of the architects on .NET and it's non-trivial to do. There's nothing actionable that can be solved in an issue like this AFAIK. The conversation here isn't talking about what would actually need to be solved to move this forward.

What is the goal here? To convince the team to investigate this work, or to flesh out the concrete investigations that would need to take place in order for this to be a reality.

**Comment by CyrusNajmabadi** [comment]
> No, it really doesn't.

It really does. You are violating rules of it, which say that it is defined in terms of threads.  It would be similar to locking a monitor on one thread, then going to another thread to release it.

>And worth restating that outside of an async context, using a Reader / Writer Lock is very simple and intuitive.

This is not about async contexts.  This is about if you use *thread* affined APIs across multiple threads.  

> Is this a bug that should be logged?

I think you're confusing async vs threads.  What you are doing would not be ok *regardless* of async/await.  If you had a readerwriterlockslim and you just did things across threads you'd have this issue as wlel.  The docs even make this clear:

> Exceptions
> SynchronizationLockException
>
> The current thread has not entered the lock in read mode.

It's necessary to abide by the rules of the underlying primitives.  If you do need to come back on another thread (surpirsing, but possible) i woudl use a rwlock that allows for that.  Note that writing such a rwlock is something that can be done.  Or, if you want to use something the framework provides, i'd recommend ConcurrentExclusiveSchedulerPair.

**Comment by CyrusNajmabadi** [comment]
Fun anecdote, my MS interview was to write a rwlock.  The one i wrote was fine to use across multiple threads :)    It didn't detect things like improper recursion.   And it was undefined what it means if you did something like release a read lock that wasn't taken, and the same with a writelock.  This was the style i learned in OS class.  Good for very few requirements.  Bad for detecting problems.

**Comment by CyrusNajmabadi** [comment]
> Should lock emit semaphoreSlim.WaitAsync instead of Montor.Enter while async?

This would likkely be a very big breaking change, and in an area that would have the chance to cause the most problems (threading primitives).  I can attest to this having very little chance of *every* happening.

What i would recommend instead would be just using the `using` construct.  I linked above how roslyn uses thsi so we can just write:

```c#
using var _ = await semaphoreSlim.WaitAsync();
// do work
```

This works nicely and even allows use of the new `using`form to prevent the need to even indent the code here if you don't want.    This sort of approach of using 'using' with a sync primitive works well, and decouples the coding patterns against a very specific clr type (like what 'lock' does).

**Comment by RobertBouillon** [comment]
@davidfowl The goal is to replace async/await. Async/await provides seamless continuation at the expense of pervasive syntax changes, pattern changes, and language restrictions. It's my goal to find a way to achieve the same benefits as async/await without the cost. I think fibers could achieve the same benefit as async/await without the restrictions it adds to the language and the runtime. This could represent the best possible change for any software application - one where the feature set remains the same, but the complexity is reduced; a net loss in lines of code without any feature loss.

The reason I believe fibers _could_ be the next evolution of asynchronous programming for C#/.NET is because the fundamental problem is a runtime problem, not a syntax problem, and this would move the solution closer to the problem. When we talk about continuation, we're talking about how to suspend a thread until it's ready to be worked by the processor again. The OS does this by managing a snapshot of the thread context (stack, registers, etc). Async/await does this by means of a state machine. Fibers would give the CLR more control over execution context, creating the opportunity to leverage different algorithms for thread suspension and continuation. 

It's risky and it's research-heavy. Nothing about this suggests it's going to be a slam dunk. But other languages are playing with it - .NET should, too. Unlike some other languages, such as Rust and Go, .NET has a Virtual Machine. For this reason, .NET may succeed where others fail (or fail where others succeed). I think an investment into researching this could yield many opportunities, even if none of those end up providing a superior alternative to async/await.

P.S. Please excuse my ignorance as I learn as I go. It seems I've been using different terms interchangeably - Green Threads, User-Mode threads, Managed Thread Scheduler, etc. I do believe the correct term I'm looking for here is **Fibers**. This pitch reflects that change to the best of my ability. Please feel free to correct any misnomers. I'll later update the title / description of this issue. Thanks @benaadams

**Comment by davidfowl** [comment]
I understand that but I'm trying to understand the purpose of this thread in relation. Is it to learn or is it just for showing interest in the idea. I think there's interest but the point being touched on in this thread are scratching the surface of what it would actually take. It's a multi-year effort to pull this off in a mature runtime that's been around for 20+ years. Not something we can patch or add a couple of APIs to make work. There are some real tough tradeoffs and considerations when trying to take .NET there. I want this experiment to happen (that's why I've been discussing it with the team) but I also understand its huge and would require a herculean effort from multiple layers of the stack. 

The need is heard and we can continue the discussion but its a big enough task and I was just trying to understand the purpose of this thread. A bulleted list of tasks and API suggestions are premature at best, there needs to be deep investigation in several areas (threading, stack management, JIT, preemption, GC, threadpool, pinning, APIs etc etc). Go had it much easier as they don't have any legacy and still they made some big tradeoffs to enable the approach (like the cost of interop). Rust doesn't have green threads and mirrors the model that .NET has. Java and Go, Erlang and Elixer (BEAM VM) are the only mainstream languages stackful user mode threads with preemption that I'm aware of.

**Comment by HaloFour** [comment]
@davidfowl 

> I want this experiment to happen (that's why I've been discussing it with the team) but I also understand its huge and would require a herculean effort from multiple layers of the stack.

I'd love to understand the thinking here.  I do imagine that it would be a herculean^2 effort to try to make the .NET runtime compatible with green threads, but much more than that is the rest of the ecosystem which would need to come along for the ride too.  With the diversity of interop the potential to block seems extremely high.  How could .NET possibly solve for that, or would the gamble be that _enough_ of the runtime itself could support green threads to make such a solution feasible?

> It's a multi-year effort to pull this off in a mature runtime that's been around for 20+ years.

Loom has been going since mid-2018 and I have my doubts that it'll even hit a preview release this year and can only assume that it'll spend at least another year in preview before going GA.

**Comment by RobertBouillon** [comment]
@davidfowl 

The purpose of the thread is determine if there's an appetite on the .NET team to pursue this endeavor, and if so, establish next steps. The scope of this project is too large for me to take on as a "pet project," and it's probably too tightly-integrated to be a bolt-on to .NET. Something like this really requires a sponsor from within the .NET team.

I see the spectrum of possible outcomes something like this:
- MS commits resources to R&D this effort
- The team sees potential value and are willing to review supporting data or efforts, but are unwilling to commit resources or define milestones
- Any investment of effort would likely be wasted as it is unlikely to be reviewed or considered 

This is certainly not a definitive list, but just to give a better idea of what I'm hoping to achieve with this thread.

Ultimately, I'd like to aim for some Proof-of-Concept which would materialize the value and the risks beyond that which analysis alone could achieve. 

MS has a lot of threading expertise inside and outside of the .NET team (OS, C++, etc), and pulling in the right people to discuss the feasibility of Fibers as a .NET feature would be more productive than an opt-in thread on GitHub, so I think we probably got a little ahead of ourselves here.

**Comment by davidfowl** [comment]
The team sees value but this is a "big rock" .NET 7/8 planning exercise at this point. I don't think we'll establish next steps on an issue like this.

**Comment by RobertBouillon** [comment]
@davidfowl If you're telling me that this is on the radar, then that's all I can really ask for and the objective of this thread has been achieved.

Is there anything I can do to get this pinned into one of those .NET 7/8 planning exercises? If not, I'm good to close the issue.

**Comment by davidfowl** [comment]
> @davidfowl If you're telling me that this is on the radar, then that's all I can really ask for and the objective of this thread has been achieved.

The current state is that we recognize there's a problem and have been having discussions about what it could look like. There's no plan beyond that. If anything this counts as an upvote in this direction!

**Comment by HaloFour** [comment]
@RobertBouillon 

> I do believe the correct term I'm looking for here is **Fibers**.

I would stick with "green threads" or "user threads".  Everyone would know what you were talking about.  Windows already has the concept of "Fibers" as application-scheduled threads but they are a different beast since they also require a stack to be preallocated for them which eliminates many of the benefits.

I'd be extremely interested in whatever findings the runtime teams might have in exploring adding green threads.  Adding the support to create, suspend and resume green threads, that is **by far** the easy part.  I'd be happy if the runtime only added that as delimited continuations would massively simplify writing coroutines.  The hard part is rewriting all of the bits of the runtime that currently block so that they don't block, because that will still block the underlying OS thread.  It also involves rewriting everything that is thread-affine so that it also understands green threads.  When a green thread is resumed it will more likely than not be on a different threadpool thread, thus existing synchronization primitives still won't work without being redesigned.  Same with anything thread local.  There would have to be a whole conversation as to what any of this means with WinForms and other UI libraries.

I find green threads to be very compelling, and I love the experience I've had with them in Go and Java Loom EA.  I don't argue against them because I don't believe they wouldn't be beneficial in the .NET world, I just can't imagine how it could be a remotely practical project.  It just requires redesigning and rewriting way too much code, much of which is outside of MS' control.  It's taken Java over 3 years to get as far as they have and they're likely still another year or two off, and their ecosystem is significantly more constrained and under their control.

**Comment by RobertBouillon** [comment]
> I would stick with "green threads" or "user threads". Everyone would know what you were talking about. 

I'm inclined to agree - I think the general idea was successfully conveyed and I don't want to get hung up on pedantic semantics :)

I agree that there are some interesting challenges to solve, as well. I'm encouraged to fork and play with it myself - given that this at least within the realm of possibilities, I won't feel like I'm wasting my time. I'd love to see what comes from a detailed design session among the vast expertise within MS. Despite my criticisms of async/await, C#'s big features have been nothing short of awe-inspiring. LINQ was the software equivalent of SpaceX landing a rocket on a boat: not only did they succeed where others failed or claimed it was too hard, they made it damned pretty. I'm reminded of LINQ's brilliance anytime I'm forced to endure Java streams. I'm excited to see where this goes, even if that means waiting a decade (hopefully not).

**Comment by vladd** [comment]
I wonder how did Java solve the following issue.

In C#, if your code is running in a single-threaded synchronization context, you can be sure that the fields are not changed between awaits. So basically you don't need any locks (as everything is anyway running on the same thread), and you need to reread the variables only after await.

Now, if we would replace async/await with green threads, basically _every_ function call would be potentially blocking and thus would be potentially yielding. This means that we would need to defensively reread the fields after every function call. (Traditional locks wouldn't help, since everything happens on the same thread in this setup.) Moreover, we would need to leave the object in a consistent state before _every_ function call. This seems to be way to restrictive.

**Comment by davidfowl** [comment]
> In C#, if your code is running in a single-threaded synchronization context, you can be sure that the fields are not changed between awaits. 

This is a very weak guarantee (nothing in the type system prevents concurrent access) that isn't hard to solve if you have a scheduler concept. Java LOOM has support for capturing continuations and it also has an [Executor](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Executor.html) concept. Continuations can be scheduled on an executor so you can have a single threaded executor (Fibers allow pluggable schedulers).

> So basically you don't need any locks (as everything is anyway running on the same thread), and you need to reread the fields only after await.

Not sure how you think async await works today but I don't see how this is different to how it works.

> Now, if we would replace async/await with green threads, basically every function call would be potentially blocking and thus would be potentially yielding. This means that we would need to defensively reread the fields after every function call. (Traditional locks wouldn't help, since everything happens on the same thread in this setup.) Moreover, we would need to bring the object into a consistent state before every function call. This seems to be way too restrictive.

I think Java bridges this gap by requiring the caller to create a virtual thread (formerly fiber) to run code. The JVM is just aware of when you're in one context or the other.

Java also doesn't have pervasive interop making this doable (though I'm sure they have issues with libraries that do JNI). It also doesn't have pinning (nor does go) which makes stack management a bit easier. There are hard problems to solve if you want magic yielding in .NET.

**Comment by RobertBouillon** [comment]
Closing to keep the backlog free of issues that are not actionable. 

Thanks to everyone who provided input. Hopefully this will make it on the roadmap, and if not, maybe we'll get to see an interesting technical write-up on why it wasn't possible.
